### ISSUE TITLE
Create a way to view the agents during training, without slowing down the fps considerably

### ISSUE THREAD

**Issue by GergelyMagyar** [MagmaMultiAgent/MagmaCore#57]
Project repo: https://github.com/MagmaMultiAgent/MagmaViz

Current design:
![asd drawio (1)](https://github.com/Getlar/VigIL-Game-Validation/assets/14542948/853c530c-da6b-4b91-824f-86429a3db3b1)

FastAPI server receiving training data (e.g. a matrix of RGB values that represent the environment) via a UDP connection and broadcasting that data to all connected users with a WebSocket.

Right now the training script creates a numpy matrix of RGB values and a simple Javascript function creates a HTML table to display the colorful 2D grid.
In the future displaying more descriptive data would be better, maybe display charts for reward or pretty much any statistic.

Constructing the data will be the train script's job. Then, with a simple imported method it could send the data to the visualization server.

The display should be completely dependent on the data received. There should be a format that the sent data must be in, and then the visualization should create charts and displays for all of them, giving the user the ability to select what to see.

The data right now is simply sent to the user and not saved anywhere, only being held in memory. This should be changed, the server should hold the received data and then the WebSocket client can request it with an API.

There should probably be some kind of persistence too, so the data won't get lost of you restart the server.

**Comment by GergelyMagyar** [comment]
Tools to make our life easier
I recommend easy to use, but powerful JS libraries

- visualization
    - [Chart.js](https://www.chartjs.org/)
- framework
    - [Svelte](https://svelte.dev/)

**Comment by GergelyMagyar** [comment]
Update: did a big refact on the project. The received data is now stored on the server and when a user selects a specific step to be displayed it sends a request to the server for the data. I also implemented a message format that allows users to display any kind of stat they want

Next step is adding the framework

**Comment by Getlar** [comment]
Completely agree with a fixed specified format for the sent data. 

I don't think there will be a huge memory issue with the user storing data in memory, however this is obviously a bad practice so your suggestion is perfect.

Also agree with having persistent data. Training takes a lot of time, hence more issues if we don't implement it. 

Suggestions that requires sending more data:

- it would be useful to see every units actions, or action distribution vector to see what the model predicts at every step.
- the same with factories
- rewards should be displayed as well, also their source (reward for digging, reward for lichen planting etc)
- theres also should be a view for every units action queue

Suggestions that do not require sending more data:

- filtering on specific agents to only see their stats
- filtering on teams sides
- having the option to go back and forth during traning between steps
- storing runs persistently, that we can go trough later in different playback speeds.

**Comment by GergelyMagyar** [comment]
Most of these are pretty much already supported (or at least theoretically possible just not fully implemented).
The format you have to send the data in is this:
```
[
	{
		episode_id: int,
		property_name: str,
		tags: list[str],
		data: Any
	},

	{
		...
	},

	...
]
```
So basically, you send a list, that contains dictionaries. Each dictionary is one specific property (some kind of statistic, e.g. the board state, the reward at that step, etc.). You can give the ID of the episode and the step, along with tags. You can put any amount of these dictionaries into one list, and you pass this list to the send method. The backend updates its stored history with the data is receives.
This is not yet implemented, but I am planning to add a feature, where the list you passed in for sending will be sliced up and put into a queue and sent separately, if the size exceeded the max UDP packet size. This way on the training side you don't really have to worry about anything, just creating the desired data.

On the frontend a visualization type is automatically applied to the data, inferred from its shape (e.g. a ndarray with size (Any, Any, 3) and integer type is an RGB matrix, so it will be displayed as such. Other visualizations are not yet implemented).

Filtering on agents, teams or pretty much anything will be possible using the tags that you can put into the dictionary.

The page is built 100% independently from the LuxAI thing, to work with any data. So it mostly comes down to what we implement on the training side.

**Comment by GergelyMagyar** [comment]
MagmaMultiAgent/MagmaViz#1
